pkg/agent/sysadvisor/plugin/qosaware/resource/memory/plugin/memory_provisioner.go

```diff
@@ -0,0 +1,124 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+
+	"github.com/kubewharf/katalyst-core/pkg/agent/qrm-plugins/memory/dynamicpolicy/memoryadvisor"
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/metacache"
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/plugin/qosaware/resource/memory/plugin/provisioner"
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/types"
+	"github.com/kubewharf/katalyst-core/pkg/config"
+	"github.com/kubewharf/katalyst-core/pkg/metaserver"
+	"github.com/kubewharf/katalyst-core/pkg/metrics"
+	"github.com/kubewharf/katalyst-core/pkg/util/general"
+	"github.com/kubewharf/katalyst-core/pkg/util/machine"
+)
+
+func init() {
+	provisioner.RegisterInitializer(types.MemoryProvisionPolicyCanonical, provisioner.NewPolicyCanonical)
+}
+
+const (
+	MemoryProvisioner = "memory-provisioner"
+)
+
+type memoryProvisioner struct {
+	mutex            sync.RWMutex
+	metaReader       metacache.MetaReader
+	metaServer       *metaserver.MetaServer
+	emitter          metrics.MetricEmitter
+	memoryProvisions machine.MemoryDetails
+	conf             *config.Configuration
+	extraConfig      interface{}
+	// policy
+	policy provisioner.ProvisionPolicy
+}
+
+func NewMemoryProvisioner(conf *config.Configuration, extraConfig interface{}, metaReader metacache.MetaReader,
+	metaServer *metaserver.MetaServer, emitter metrics.MetricEmitter) MemoryAdvisorPlugin {
+	mp := &memoryProvisioner{
+		conf:             conf,
+		extraConfig:      extraConfig,
+		metaReader:       metaReader,
+		metaServer:       metaServer,
+		emitter:          emitter,
+		memoryProvisions: make(machine.MemoryDetails),
+	}
+	klog.Infof("[memory-provisioner] new memory provisioner")
+	if err := mp.initializeMemoryProvisioner(); err != nil {
+		general.Errorf("[memory-provisioner] initialize memory provisioner failed: %v", err)
+		return nil
+	}
+
+	return mp
+}
+
+// initializeMemoryProvisioner initialize memory provisioner
+func (m *memoryProvisioner) initializeMemoryProvisioner() error {
+	// get policy
+	policyName := m.conf.MemoryProvisionPolicy
+	initializers := provisioner.GetRegisteredInitializers()
+	reservedForAllocate := m.conf.GetDynamicConfiguration().
+		ReservedResourceForAllocate[v1.ResourceMemory]
+	initFunc, ok := initializers[policyName]
+	klog.Infof("[memory-provisioner] policyName: %s", policyName)
+	if !ok {
+		return fmt.Errorf("policy %s not found", policyName)
+	}
+	m.policy = initFunc(m.conf, nil, m.metaReader, m.metaServer, m.emitter)
+	m.policy.SetEssentials(
+		types.ResourceEssentials{
```

Comment:
set essential in Reconcile to make dynamicConfiguration available, because initializeMemoryProvisioner only run once

Reply:
done